### PR TITLE
Add 'info' diagnostics

### LINKED ISSUE #19392: codeFixes for things that aren't diagnostics
**TypeScript Version:**  2.7.0-dev.20171020

**Code**

```ts
function f(x) {}
```

It would be nice to have a codeFix to fix implicit-any instead of a refactoring. I believe a refactoring requires you to highlight the thing in question, which isn't easily discoverable.
Currently codeFixes work by iterating over errors; but in this case we would want to produce one even in an error-less build.
This would also help us avoid having to create codeFix / refactor pairs for things such as installing types for implicit-any packages.
We could also generate "info" errors (ref: #14501) that don't surface normally and make codeFixes for those.

CC @mjbvz

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #19392

Previously we had just syntax and semantic diagnostics. This adds a third layer of "info" diagnostics. They are only a part of services, so do not impact command line compilation.

Now when the editor sends us a "geterr" request, we will send *3* responses for "syntaxDiag", "semanticDiag", and "infoDiag". (See changes to `updateErrorCheck`.) All return a Diagnostic[].

For use from Visual Studio, this also adds a "infoDiagnosticsSync" command. (@amcasey or would it be easier to just send it along with "semanticDiagnosticsSync"? Might impact performance as we add more info diagnostics though.)

For now there is only one info diagnostic, for a file that could be converted to commonJs. Moved the `convertToEs6Module` refactoring to a codefix.

CC @mjbvz

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -4017,8 +4017,14 @@ namespace ts {
     export enum DiagnosticCategory {
         Warning,
         Error,
+        Info,
```
Comment (mhegazy) [inline src/compiler/types.ts]: I would call it `Suggession`. `Info` and `Message` are pretty much the same thing.

File: src/server/protocol.ts [hunk 2]
```diff
@@ -2121,7 +2135,7 @@ namespace ts.server.protocol {
         text: string;
 
         /**
-         * The category of the diagnostic message, e.g. "error" vs. "warning"
+         * The category of the diagnostic message. This is the lower-case of a DiagnosticCategory member.
```
Comment (mhegazy) [inline src/server/protocol.ts]: the generated protocol.d.ts does not have `DiagnosticCategory`.  so I would keep the comment in there. 

One other thing also is we allow extensions  to set it, so it is possible that there are other values here. so a union of string literals would be too restrictive for clients.

File: src/server/session.ts [hunk 3]
```diff
@@ -482,38 +479,64 @@ namespace ts.server {
 
         private syntacticCheck(file: NormalizedPath, project: Project) {
             try {
-                const diags = project.getLanguageService().getSyntacticDiagnostics(file);
-                if (diags) {
-                    const bakedDiags = diags.map((diag) => formatDiag(file, project, diag));
-                    this.event<protocol.DiagnosticEventBody>({ file, diagnostics: bakedDiags }, "syntaxDiag");
+                // TODO: why do we check for diagnostics existence here, but in semanticCheck send unconditionally?
```
Comment (mhegazy) [inline src/server/session.ts]: we should report both unconditionally

File: src/services/infoDiagnostics.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+/* @internal */
+namespace ts {
+    export function infoDiagnostics(sourceFile: SourceFile): Diagnostic[] {
```
Comment (mhegazy) [inline src/services/infoDiagnostics.ts]: consider naming it `compute*Diagnostics`

**Comment by mhegazy** [comment]
Just a few comments about the name of the enum value. @amcasey any comments?

File: src/compiler/types.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/types.ts]: Suggestion

Comment (amcasey) [reply src/compiler/types.ts]: FWIW, Roslyn calls it `Info`.  What *is* the difference between `Info` and `Message`?  Is `Message` actually `Telemetry`?

Comment (mhegazy) [reply src/compiler/types.ts]: Message is like description of a quick fix. user never sees that as a Diagnostic really.

Comment (amcasey) [reply src/compiler/types.ts]: This isn't quite what Roslyn calls `Info`, so I'm fine with either `Info` or `Suggestion` (or `Hint`, which appears to be the VS Code name).

File: src/server/protocol.ts [hunk 5]
```diff
@@ -2010,6 +2011,19 @@ namespace ts.server.protocol {
         body?: Diagnostic[] | DiagnosticWithLinePosition[];
     }
 
+    export interface InfoDiagnosticsSyncRequest extends FileRequest {
+        command: CommandTypes.InfoDiagnosticsSync;
+        arguments: InfoDiagnosticsSyncRequestArgs;
+    }
+
+    export interface InfoDiagnosticsSyncRequestArgs extends FileRequestArgs {
```
Comment (amcasey) [inline src/server/protocol.ts]: Can/should interface declarations be shared across the various kinds of diagnostics?

File: src/server/session.ts (hunk 3 above)
Comment (amcasey) [reply src/server/session.ts]: FYI @aozgaa

File: src/services/codefixes/convertToEs6Module.ts [hunk 6]
```diff
@@ -1,85 +1,23 @@
...
-        const edits = textChanges.ChangeTracker.with(context, changes => {
-            const moduleExportsChangedToDefault = convertFileToEs6Module(file, program.getTypeChecker(), changes, program.getCompilerOptions().target);
-            if (moduleExportsChangedToDefault) {
-                for (const importingFile of program.getSourceFiles()) {
-                    fixImportOfModuleExports(importingFile, file, changes);
+namespace ts.codefix {
+    const errorCodes = [Diagnostics.File_is_a_CommonJS_module_it_may_be_converted_to_an_ES6_module.code];
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
```
Comment (amcasey) [inline src/services/codefixes/convertToEs6Module.ts]: This seems much simpler. :smile:

File: tests/cases/fourslash/javascriptModules22.ts [hunk 7]
```diff
@@ -29,4 +29,4 @@ verify.completionListContains("name");
 edit.insert("name;\nsausages.");
 verify.completionListContains("eggs");
 edit.insert("eggs;");
-verify.noErrors();
+verify.noErrors({ ignoreInfoDiagnostics: true });
```
Comment (amcasey) [inline tests/cases/fourslash/javascriptModules22.ts]: Personally, I would expect `noDiagnostics` to mean "no diagnostics of any severity" and `noErrors` to mean "no diagnostics of error severity".

**Comment by amcasey** [comment]
I've created an internal PR that makes VS consume the new API.  LGTM

File: src/compiler/types.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/types.ts]: I'd rather be consistent, but if we plan to have more-specific categories than `Info` in the future, `Suggestion` is fine with me.

**Comment by andy-ms** [comment]
@amcasey @mhegazy good to go? (Note that since name has changed to "suggestion", VS will need to change too.)

**Comment by amcasey** [comment]
I'm fine with either "Info" or "Suggestion" and I've already confirmed that I can consume the new API from VS.

**Pull request merged**